src/cached_object_store.rs

```diff
@@ -762,6 +797,224 @@ impl LocalCacheEntry for FsCacheEntry {
+            }
+            Ok(()) => {
+                debug!(
+                    "evictor: evicted cache file: {:?}, bytes: {}",
+                    target, target_bytes
+                );
+                target_bytes
+            }
+        }
+    }
+
+    pub async fn scan_cache_folder_bytes(
+        root_folder: &std::path::PathBuf,
+    ) -> Result<usize, Box<dyn std::error::Error>> {
+        let mut total_bytes = 0;
+        let iter = WalkDir::new(root_folder).into_iter();
+        for entry in iter {
+            let entry = entry?;
+            let metadata = tokio::fs::metadata(entry.path()).await?;
+            total_bytes += metadata.len();
+        }
+        Ok(total_bytes as usize)
+    }
+
+    // pick a file to evict, return None if no file is picked. it takes a pick-of-2 strategy, randomized pick
+    // two of files, compare their last access time, and evict the older one.
+    async fn pick_evict_target(&mut self) -> Option<(std::path::PathBuf, u64)> {
+        // extend the evict_buffer if it's empty. it will randomized iterate the cache folder,
+        // and randomly pick 1000 files into the evict_buffer.
+        let mut rng = rand::rngs::StdRng::from_entropy();
+        if self.evict_buffer.len() < 2 {
+            let iter = WalkDir::new(&self.root_folder).into_iter();
+            for entry in iter.choose_multiple(&mut rng, 1000) {
+                let entry = match entry {
+                    Ok(entry) => entry,
+                    Err(err) => {
+                        warn!("evictor: unexpected error on reading dir entry to pick the evict target, skip the entry: {}", err);
+                        break;
+                    }
+                };
+
+                if entry.file_type().is_dir() {
+                    continue;
+                }
+                self.evict_buffer.push_back(entry.path().to_path_buf());
+            }
+        }
+
+        // if the evict_buffer length still below 2, return None
+        if self.evict_buffer.len() < 2 {
+            return None;
+        }
+
+        // pop two files from the evict_buffer
+        let (path0, path1) = {
+            match (self.evict_buffer.pop_front(), self.evict_buffer.pop_front()) {
+                (Some(path0), Some(path1)) => (path0, path1),
+                _ => return None,
+            }
+        };
+
+        // read the accessed time & length of the two files
+        let (metadata0, metadata1) = match (
+            tokio::fs::metadata(&path0).await,
+            tokio::fs::metadata(&path1).await,
+        ) {
+            (Ok(metadata0), Ok(metadata1)) => (metadata0, metadata1),
+            (Err(err), _) | (_, Err(err)) => {
+                warn!(
+                    "evictor: failed to read metadata of the cache files: {}",
+                    err
+                );
+                return None;
+            }
+        };
+        let (accessed0, accessed1) = match (metadata0.accessed(), metadata1.accessed()) {
```

Comment:
I think most linux distros default fs mounts to turn on the `noatime` or `relatime` options to avoid doing a write to update the file metadata on every read. With `noatime` the access time is never updated. With `relatime` its updated whenever the file is read after its modified or its been longer than 24 hours (by default) since it was last accessed. If these are set then this won't really approximate LRU (it just becomes random eviction). We should probably at least document that and recommend users to mount with `atime` enabled. The caveat then is that this will likely hurt performance, so we should try to measure that. I'm not sure if you can change the interval used by relatime - that would be one way to optimize. Alternatively we can track access times in memory if it's a problem.

Reply:
oh, this is a big issue according to the design. I haven't noticed it before. I need to do some further investigation into what Alluxio/JuiceFS does in this case.

it's a big risk. i'll convert this PR to draft until more investigation is completed.

thank you for the information!

Reply:
Why do they rescan every hour? 🤔

Reply:
it feels some kind of eventual consistency for me 🤔, in my understanding the in-memory map does not really requires strong consistency with the disk files, but just to help finding the files to evict. if the in-memory map for any reason lost some entries during updates, the map could be finally repaired in the rescan phase 😵.

Reply:
🤔 I kinda feel like we should just KISS and use an in-memory map to track last access. On startup, it's blank and we just nuke stuff at random if neither 2-choice object has a last access time.

Reply:
+1 :+1:, let me do this